x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexerTests.java

```diff
@@ -382,22 +457,135 @@ public void testStop_WhileIndexing() throws Exception {
+        doTestFiveRuns(18_000, Collections.emptyList());
+    }
+
+    public void testFiveRunsThrottled1000000() throws Exception {
+        // request per seconds is set high, so throttling does not kick in
+        doTestFiveRuns(1_000_000, Collections.emptyList());
+    }
+
+    public void doTestFiveRuns(float requests_per_second, Collection<TimeValue> expectedDelays) throws Exception {
         AtomicReference<IndexerState> state = new AtomicReference<>(IndexerState.STOPPED);
-        final ExecutorService executor = Executors.newFixedThreadPool(1);
+        final MockThreadPool threadPool = new MockThreadPool(getTestName());
         try {
-            MockIndexerFiveRuns indexer = new MockIndexerFiveRuns (executor, state, 2);
+            MockIndexerFiveRuns indexer = new MockIndexerFiveRuns (threadPool, ThreadPool.Names.GENERIC, state, 2, requests_per_second,
+                null);
             indexer.start();
             assertThat(indexer.getState(), equalTo(IndexerState.STARTED));
             assertTrue(indexer.maybeTriggerAsyncJob(System.currentTimeMillis()));
             assertBusy(() -> assertTrue(isFinished.get()));
             indexer.assertCounters();
+            threadPool.assertCountersAndDelay(expectedDelays);
         } finally {
-            executor.shutdownNow();
+            ThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
         }
     }
+
+    public void testFiveRunsRethrottle0_100() throws Exception {
+        doTestFiveRunsRethrottle(-1, 100, timeValueCollectionFromMilliseconds(9950L));
+    }
+
+    public void testFiveRunsRethrottle100_0() throws Exception {
+        doTestFiveRunsRethrottle(100, 0, timeValueCollectionFromMilliseconds(9950L, 9950L, 9950L));
+    }
+
+    public void testFiveRunsRethrottle100_1000() throws Exception {
+        doTestFiveRunsRethrottle(100, 1000, timeValueCollectionFromMilliseconds(9950L, 9950L, 9950L, 950L));
+    }
+
+    public void testFiveRunsRethrottle1000_100() throws Exception {
+        doTestFiveRunsRethrottle(1000, 100, timeValueCollectionFromMilliseconds(950L, 950L, 950L, 9950L));
+    }
+
+    public void doTestFiveRunsRethrottle(
+        float requests_per_second,
+        float requests_per_second_rethrottle,
+        Collection<TimeValue> expectedDelays
+    ) throws Exception {
+        AtomicReference<IndexerState> state = new AtomicReference<>(IndexerState.STOPPED);
+
+        final MockThreadPool threadPool = new MockThreadPool(getTestName());
+        try {
+            CountDownLatch latch = new CountDownLatch(1);
+            MockIndexerFiveRuns indexer = new MockIndexerFiveRuns (threadPool, ThreadPool.Names.GENERIC, state, 2, requests_per_second,
+                latch);
+            indexer.start();
+            assertThat(indexer.getState(), equalTo(IndexerState.STARTED));
+            assertTrue(indexer.maybeTriggerAsyncJob(System.currentTimeMillis()));
+            // wait until the indexer reached latch await
```

Comment:
This comment initially confused me.  It might be clear to say:

```suggestion
            // wait until the indexer starts waiting on the latch
```